### ISSUE TITLE
Dependent Types: no way to differentiate a field that hides a superclass field

### ISSUE THREAD

**Issue by GoogleCodeExporter** [typetools/checker-framework#152]
```
Attached is a simple program that will exercise this bug.  There are comments 
describing the bug in the code.

Short Description:  If a class and its base class both have a member "lock" 
and fields in both the class and base class are guarded by their respective 
locks then the checker does not report an error when the base class's guarded 
members are synchronized on the subclasses "lock" field even though the 
subclasses "lock" field is a different object from the base classes "lock" 
field.

To Reproduce:
1.  Download the attached file(LockBugExample.java)
2.  Run javac -processor checkers.lock.LockChecker LockBugExample.java

Expected Output:
Unguarded access errors for lines:  45,55,56,58

Actual Output:
A fully compiled program
```

Original issue reported on code.google.com by `Jonathan...@gmail.com` on 28 Aug 2012 at 10:43

Attachments:
- [LockBugExample.java](https://storage.googleapis.com/google-code-attachments/checker-framework/issue-152/comment-0/LockBugExample.java)

**Comment by smillst** [comment]
@jthaine  Could you please add this test case to the repository?

**Comment by smillst** [comment]
Should be fixed once #770 is fixed.

**Comment by mernst** [comment]
Issue #770 is fixed, but this issue remains a problem.  The two tests for issue #152 remain broken (they are currently disabled).

**Comment by smillst** [comment]
framework/tests/flowexpression/ThisSuper.java is another test case.

**Comment by HuM4NoiD** [comment]
I wanted to ask, does it affect other tests, if instead to just using `this`, we can use `MySubClass.this`? That is while converting `ThisReference` to a string using `toString()`, we just return "this" and thus the checker cannot differentiate between `MyBaseClass` and `MySubClass extends MyBaseClass`. If we do this, then the errors are reported in the required places for the class `$CHECKERFRAMEWORK/checker/tests/lock` Let me know it doing this is a hack or it can work.

**Comment by mernst** [comment]
I cannot follow your question.  Can you create a test case that shows the exact code you have in mind (it sounds like there are two variants), and also show the checker's output on each of the variants?  Or if you are talking about a change to the checker, show the exact code changes in the Checker Framework.

**Comment by HuM4NoiD** [comment]
@mernst I have created Pull request in my own fork [here](https://github.com/HuM4NoiD/checker-framework/pull/3), I have explained the fix
